File: python/p049.py
# 
# Solution to Project Euler problem 49
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 

import eulerlib


def compute():
	LIMIT = 10000
	isprime = eulerlib.list_primality(LIMIT - 1)
	for base in range(1000, LIMIT):
		if isprime[base]:
			for step in range(1, LIMIT):
				a = base + step
				b = a + step
				if     a < LIMIT and isprime[a] and has_same_digits(a, base) \
				   and b < LIMIT and isprime[b] and has_same_digits(b, base) \
				   and (base != 1487 or a != 4817):
					return str(base) + str(a) + str(b)
	raise RuntimeError("Not found")


def has_same_digits(x, y):
	return sorted(str(x)) == sorted(str(y))


if __name__ == "__main__":
	print(compute())

File: java/p049.java
/* 
 * Solution to Project Euler problem 49
 * Copyright (c) Project Nayuki. All rights reserved.
 * 
 * https://www.nayuki.io/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */

import java.util.Arrays;


public final class p049 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p049().run());
	}
	
	
	private static final int LIMIT = 10000;
	
	public String run() {
		boolean[] isPrime = Library.listPrimality(LIMIT - 1);
		for (int base = 1000; base < LIMIT; base++) {
			if (isPrime[base]) {
				for (int step = 1; step < LIMIT; step++) {
					int a = base + step;
					int b = a    + step;
					if (       a < LIMIT && isPrime[a] && hasSameDigits(a, base)
					        && b < LIMIT && isPrime[b] && hasSameDigits(b, base)
					        && (base != 1487 || a != 4817))
						return "" + base + a + b;
				}
			}
		}
		throw new RuntimeException("Not found");
	}
	
	
	private static boolean hasSameDigits(int x, int y) {
		char[] xdigits = Integer.toString(x).toCharArray();
		char[] ydigits = Integer.toString(y).toCharArray();
		Arrays.sort(xdigits);
		Arrays.sort(ydigits);
		return Arrays.equals(xdigits, ydigits);
	}
	
}
